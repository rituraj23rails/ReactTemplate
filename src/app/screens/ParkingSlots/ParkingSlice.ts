/**
 PARKING SLOTS SLICE
 */
 import { createSlice, PayloadAction } from '@reduxjs/toolkit'

 export interface LOCATION_STATE {
   parkingData: Array<any>;
 }
 
 const initialState: LOCATION_STATE = {
   parkingData: []
 }
 
 /* istanbul ignore next */
 export const parkingSlice = createSlice({
   name: 'Parking',
   initialState,
   reducers: {
     updateParkingData: (state, action) => {
        state.parkingData = action.payload;
     }
   },
 })
 
 // Action creators are generated for each case reducer function
 export const { updateParkingData } = parkingSlice.actions
 
 export default parkingSlice.reducer;