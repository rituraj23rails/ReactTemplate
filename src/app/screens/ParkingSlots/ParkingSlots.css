.slotMainContainer {
    display: flex;
    flex-direction: column;
    align-items: center;
}

.listContainer {
    width: 100%;
    display: flex;
    flex-direction: row;
    flex-wrap: wrap;
    justify-content: space-between;
    padding: 50px 50px;
}

.slotCell {
    width: 150px;
    height: 200px;
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    border-radius: 2px;
    margin: 0px auto 20px auto;
}

.availableText {
    font-size: 20px;
    color: black;
    font-weight: bold;
}

.regisrationTextNo {
    font-size: 20px;
    color: white;
    margin-top: 20px;
}

.textCarInputField {
    width: 300px;
    height: 50px;
    border: 2px solid #000;
    border-radius: 2px;
    font-size: 20px;
    padding-left: 10px;
}

.textCarInputField:focus {
    outline: none;
}

.submitCarButton {
    cursor: pointer;
    width: 200px;
    height: 50px;
    display: flex;
    justify-content: center;
    align-items: center;
    border: 2px solid #000;
    margin-top: 50px;
    margin-bottom: 50px;
}

.submitCarButton:hover {
    background-color: green;
    color: #fff;
}




