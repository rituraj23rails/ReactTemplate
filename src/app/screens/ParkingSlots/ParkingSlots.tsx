import React, { Component, useRef } from 'react';
import './ParkingSlots.css';
import WithRouter from '../../../WithRouter';
import { connect } from 'react-redux';
import { updateParkingData } from './ParkingSlice';

export interface Props {
    navigate:  any;
    id: string;
    parkingData: Array<any>;
    location: any;
    updateParkingData: (parkingData: any) => void;
}

interface S {
    parkingSlots: string;
    parkingData: Array<any>;
    carRegistrationNo: string;
}

interface SS {
    id: any;
}

class ParkingSlots extends Component<Props, S, SS> {
    carousel: any = null;
    constructor(props: any) {
        super(props);
        this.state = {
            parkingSlots: this.props.location.state.parkingSlots,
            parkingData: [],
            carRegistrationNo: ''
        };
    };

    componentDidMount = () => {
        if(this.props.parkingData && this.props.parkingData.length === 0) {
            let parkingData: Array<any> = [];
            for(let i = 0; i < Number(this.state.parkingSlots); i++) {
                parkingData.push({
                    isAvailable: true,
                    carRegistrationNo: '',
                    uniqueNo: i
                })
            }
            this.setState({ parkingData: parkingData }, () => {
            });
        } else {
            this.setState({ parkingData: this.props.parkingData });
        }
    }

    componentWillUnmount = () => {

    }

    carRegistration = () => {
        if (this.state.carRegistrationNo.trim().length === 0) {
            alert('Car Registration No is empty.');
            return;
        }
        let availableParkingArea = this.state.parkingData;
        let availableRegistrationIndex = availableParkingArea.findIndex((item: any) => item.carRegistrationNo.toLowerCase() === this.state.carRegistrationNo.toLowerCase());
        if(availableRegistrationIndex !== -1) {
            alert('Already Car Registered with this No.');
            return;
        }
        let availableIndex = availableParkingArea.findIndex((item: any) => item.isAvailable === true);
        if (availableIndex === -1) {
            alert('Parking is full.');
            return;
        }
        let availableIndexes: any = [];
        availableParkingArea.map((item: any, index: any) => {
            if (item.isAvailable) {
                availableIndexes.push(index);
            }
        });
        let randomIndex = Math.floor((Math.random() * availableIndexes.length));
        availableParkingArea[availableIndexes[randomIndex]].isAvailable = false;
        availableParkingArea[availableIndexes[randomIndex]].carRegistrationNo = this.state.carRegistrationNo;
        availableParkingArea[availableIndexes[randomIndex]].time = new Date().getTime();
        this.setState({ parkingData: availableParkingArea, carRegistrationNo: '' }, () => {
            this.props.updateParkingData(JSON.parse(JSON.stringify(this.state.parkingData)));
            alert('Car Registered Successfully');
        });
    }

    handleChangeTextInput = (event: any) => {
        event?.preventDefault();
        this.setState({ carRegistrationNo: event.target.value });
    }

    onPressSlot = (item: any) => {
        this.props.navigate(`/parkingCharges`, { state: { slotData: item }});
    }

    renderSlotsList = () => {
        return (
            <div className="listContainer">
                {this.state.parkingData.map((item: any, index: any) => {
                    return (
                        <div data-testid="cellButtonTestID" onClick={() => !item.isAvailable && this.onPressSlot(item)} key={index} className="slotCell" style={{ backgroundColor: item.isAvailable ? 'green' : 'red' }}>
                            <p className="availableText">{item.isAvailable ? 'Available' : 'Not Available'}</p>
                            <p className="regisrationTextNo">{item.carRegistrationNo}</p>
                        </div>
                    )
                })}
            </div>
        )
    }

    render() {
        return (
            <div className='slotMainContainer'>
                {this.renderSlotsList()}
                <input data-testid="textInputTestID" type="text" placeholder='Enter Car Registration No' className='textCarInputField' value={this.state.carRegistrationNo} onChange={(event) => this.handleChangeTextInput(event)} />
                <button data-testid="submitButtonTestID" onClick={() => this.carRegistration()} className="submitCarButton">CAR REGISTRATION</button>
            </div>
        )
    }
};

const mapStateToProps = (state: any) => ({
    parkingData: state.parking.parkingData
});

const mapDispatchToProps = {
    updateParkingData: updateParkingData
};

export default connect(mapStateToProps, mapDispatchToProps)(WithRouter(ParkingSlots));
export {ParkingSlots};
