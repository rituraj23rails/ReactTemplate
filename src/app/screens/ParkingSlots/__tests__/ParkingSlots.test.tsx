/**
 *
 PARKING HOME TEST FILE
 * @format
 */
import React from 'react';
import renderer from 'react-test-renderer';
import { shallow, ShallowWrapper, configure, render } from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import { ParkingSlots } from '../ParkingSlots';

configure({ adapter: new Adapter() });


const screenProps = {
    navigate: jest.fn(),
    id: 'ParkingSlots',
    parkingData: [],
    location: {
        state: {
            parkingSlots: 2
        }
    },
    updateParkingData: jest.fn()
}

const screenProps1 = {
    navigate: jest.fn(),
    id: 'ParkingSlots',
    parkingData: [{
        isAvailable: true,
        carRegistrationNo: '123',
        uniqueNo: 1
    }],
    location: {
        state: {
            parkingSlots: 0
        }
    },
    updateParkingData: jest.fn()
}

const screenProps2 = {
    navigate: jest.fn(),
    id: 'ParkingSlots',
    parkingData: [],
    location: {
        state: {
            parkingSlots: 0
        }
    },
    updateParkingData: jest.fn()
}

const screenProps3 = {
    navigate: jest.fn(),
    id: 'ParkingSlots',
    parkingData: [{
        isAvailable: false,
        carRegistrationNo: '111',
        uniqueNo: 1
    }],
    location: {
        state: {
            parkingSlots: 0
        }
    },
    updateParkingData: jest.fn()
}

const screenProps4 = {
    navigate: jest.fn(),
    id: 'ParkingSlots',
    parkingData: [{
        isAvailable: true,
        carRegistrationNo: '',
        uniqueNo: 3
    }],
    location: {
        state: {
            parkingSlots: 0
        }
    },
    updateParkingData: jest.fn()
}

describe('Launching ParkingSlots', () => {
    
    beforeEach(() => {
        global.fetch = jest.fn()
        window.alert = jest.fn();
    })

    it('ParkingSlots root file loads properly', () => {
        const tree: any = renderer.create(<ParkingSlots {...screenProps} />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('Test User ParkingSlots Empty Data Screen', () => {
        let wrapper: ShallowWrapper;
        let instance: ParkingSlots;
        wrapper = shallow(<ParkingSlots {...screenProps} />)
        instance = wrapper.instance() as ParkingSlots;
        let inputComponent = wrapper.findWhere((node) => node.prop("data-testid") === "textInputTestID");
        inputComponent.simulate('change', { preventDefault: jest.fn(), target: { value: 'ind' } });
        let buttonComponent = wrapper.findWhere((node) => node.prop("data-testid") === "submitButtonTestID");
        buttonComponent.simulate("click");
        instance.componentWillUnmount();
    });

    it('Test User ParkingSlots Data Screen', () => {
        let wrapper: ShallowWrapper;
        let instance: ParkingSlots;
        wrapper = shallow(<ParkingSlots {...screenProps1} />)
        instance = wrapper.instance() as ParkingSlots;
        instance.setState({ carRegistrationNo: '' }, () => {
            instance.carRegistration();
        })
        instance.setState({ carRegistrationNo: '123' }, () => {
            instance.carRegistration();
        })
        instance.componentWillUnmount();
    });

    it('Test User ParkingSlots non empty parking Slot', () => {
        let wrapper: ShallowWrapper;
        let instance: ParkingSlots;
        wrapper = shallow(<ParkingSlots {...screenProps2} />)
        instance = wrapper.instance() as ParkingSlots;
        instance.setState({ carRegistrationNo: '233' }, () => {
            instance.carRegistration();
        })
        instance.componentWillUnmount();
    });

    it('Test User ParkingSlots Cell', () => {
        let wrapper: ShallowWrapper;
        let instance: ParkingSlots;
        wrapper = shallow(<ParkingSlots {...screenProps3} />)
        instance = wrapper.instance() as ParkingSlots;
        instance.state.parkingData.forEach((item:any) => {
            let buttonComponent = wrapper.findWhere((node) => node.prop("data-testid") === "cellButtonTestID");
            buttonComponent.simulate("click");
        })
        instance.carRegistration();
        instance.componentWillUnmount();
    });

    it('Test User Available ParkingSlots Cell', () => {
        let wrapper: ShallowWrapper;
        let instance: ParkingSlots;
        wrapper = shallow(<ParkingSlots {...screenProps4} />)
        instance = wrapper.instance() as ParkingSlots;
        instance.setState({ carRegistrationNo: '233' }, () => {
            instance.carRegistration();
        })
        instance.componentWillUnmount();
    });
});