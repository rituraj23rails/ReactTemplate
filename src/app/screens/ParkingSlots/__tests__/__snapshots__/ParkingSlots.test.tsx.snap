// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Launching ParkingSlots ParkingSlots root file loads properly 1`] = `
<div
  className="slotMainContainer"
>
  <div
    className="listContainer"
  >
    <div
      className="slotCell"
      data-testid="cellButtonTestID"
      onClick={[Function]}
      style={
        Object {
          "backgroundColor": "green",
        }
      }
    >
      <p
        className="availableText"
      >
        Available
      </p>
      <p
        className="regisrationTextNo"
      />
    </div>
    <div
      className="slotCell"
      data-testid="cellButtonTestID"
      onClick={[Function]}
      style={
        Object {
          "backgroundColor": "green",
        }
      }
    >
      <p
        className="availableText"
      >
        Available
      </p>
      <p
        className="regisrationTextNo"
      />
    </div>
  </div>
  <input
    className="textCarInputField"
    data-testid="textInputTestID"
    onChange={[Function]}
    placeholder="Enter Car Registration No"
    type="text"
    value=""
  />
  <button
    className="submitCarButton"
    data-testid="submitButtonTestID"
    onClick={[Function]}
  >
    CAR REGISTRATION
  </button>
</div>
`;
